YunoHost/yunohost
# Misc optimizations to speed up regen-conf and other things

## The problem

Dear journal, here is today's story of "I got mad because of [insert stupid issue] and made yet another PR"

Today I was helping nicofrand with some regenconf issues (probably related to  PR 946), told him to `regen-conf nginx --dry-run` (just nginx!).  After a couple of minutes I get impatient about knowing the results yet he tells me that the command was still running as if it was expected that it takes long, supposedly because the server host "many" domains (well, 10, imho that ain't so many). It ended up taking something like a whooping *5 minutes* in total, again, just for nginx.

His server clearly had some memory issues, yet this is not uncommon to see these commands taking so much time. Tested on my RPi2 that has just 2 domain and the command still takes a good 30ish second for a `regen-conf nginx` to complete, and 60ish second for the full regenconf. This is a crazy amount of time considering what the regen-conf does

Investigating I noticed that : 
- Calling `yunohost domain list` inside the regen-conf hooks take a huge lot of time, about 3ish second per call, 
- Inside nginx conf regen, calling `yunohost domain cert-status` takes 10ish seconds (TEN!) ... Probably related to the fact that it has to load the `requests` module and does a bunch of HTTP GET ... (Similarly, see the time that `yunohost domain dns-conf` takes...)
- And all this make me saw once again how crazy some things are about fetching the IP from ip.yunohost.org *all the time*. Like `yunohost domain cert-status` will fetch the IP once per domain (yes, inside the same command)
- There are other calls like `yunohost setting get` that takes 3ish seconds just to fetch a value and slow things as well, but too lazy tonight to speed that up (I think it's a bit more complicated than just reading a yaml file)

## Solution

- Feed the domain list to conf regen hooks via bash env vars to avoid running `yunohost domain list`
- Call `yunohost domain cert-status` a single time instead of calling it for each domain
- Add a cache mechanism to `get_public_ip`

Many things could still be improved, but on a slow-hardware 10-domain setup that should easily save a good one or two minutes when running the regenconf, which ain't nothing... Also `domain dns-conf` goes from 10ish secs to 2.3 secs if the IP cache is still valid...

## PR Status

Tested and working on my side

## How to test

Run `domain dns-conf`,  `domain cert-status`, `tools regen-conf nginx`, etc... in debug mode, ideally on slow hardware (but even on fast hardware you'll see some improvements..)

## Validation
